### PR TITLE
cluster: fix some issues of tidb cluster with tispark component

### PR DESCRIPTION
**Pull request by AstroProfundis**
### What problem does this PR solve? 
Fix some issues found during test.

### What is changed and how it works?
 - Fix component role handling, `-R` in sub commands now works properly on tispark nodes
 - Fix wrong validation tag for tispark config fields
 - Fix listening address handling for tispark workers

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

Release notes:

```release-note
NONE
```

### CODE REVIEW

File: pkg/cluster/spec/spec.go [hunk 1]
```diff
@@ -202,7 +202,12 @@ func (s *Specification) BaseTopo() *BaseTopo {
 func AllComponentNames() (roles []string) {
 	tp := &Specification{}
 	tp.IterComponent(func(c Component) {
-		roles = append(roles, c.Name())
+		switch c.Name() {
+		case ComponentTiSpark: // tispark-{master, worker}
```
Comment (lonng) [inline pkg/cluster/spec/spec.go:206]: I think we should treat `tispark-{master,worker}` as a different component type.

Comment (lonng) [reply pkg/cluster/spec/spec.go:206]: And we don't need `Role` method anymore.

Comment (AstroProfundis) [reply pkg/cluster/spec/spec.go:206]: We need something called `tispark` to download the package

File: pkg/cluster/spec/tispark.go [hunk 2]
```diff
@@ -39,8 +39,8 @@ type TiSparkMasterSpec struct {
 	WebPort      int                    `yaml:"web_port" default:"8080"`
 	DeployDir    string                 `yaml:"deploy_dir,omitempty"`
 	JavaHome     string                 `yaml:"java_home,omitempty" validate:"java_home:editable"`
-	SparkConfigs map[string]interface{} `yaml:"spark_config,omitempty" validate:"spark_config:editable"`
-	SparkEnvs    map[string]string      `yaml:"spark_env,omitempty" validate:"spark_env:editable"`
+	SparkConfigs map[string]interface{} `yaml:"spark_config,omitempty" validate:"spark_config:ignore"`
```
Comment (lonng) [inline pkg/cluster/spec/tispark.go:42]:
```suggestion
	SparkConfigs map[string]interface{} `yaml:"spark_config,omitempty" validate:"-"`
```

File: pkg/cluster/spec/tispark.go [hunk 3]
```diff
@@ -39,8 +39,8 @@ type TiSparkMasterSpec struct {
...
+	SparkEnvs    map[string]string      `yaml:"spark_env,omitempty" validate:"spark_env:ignore"`
```
Comment (lonng) [inline pkg/cluster/spec/tispark.go:43]: ditto

File: pkg/cluster/spec/tispark.go (hunk 2 above)
Comment (AstroProfundis) [reply pkg/cluster/spec/tispark.go:42]: `-` does not make any sense. The `validate` tag is handled in `pkg/utils/diff.go` and only `ignore` makes the content of these config fields freely editable.

**Review by lonng — APPROVED** [review]
LGTM
But I think the `Instance` interface is getting bigger, we should simplify it.

**Pull request merged**